include/pybind11/detail/class.h

```diff
@@ -268,7 +282,9 @@ inline PyTypeObject *make_default_metaclass() {
     type->tp_base = type_incref(&PyType_Type);
     type->tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HEAPTYPE;
 
+#if defined(PYPY_VERSION)
```

Comment:
The new mechanism does not need to override tp_call?

Reply:
I think what is missing in the context is that pybind11_meta_call only conducts protection for base class init, and can thus be skipped on CPYTHON.

Maybe defining a few secondary compilation conditions,

```
#if defined(PYPY_VERSION)
#define PYBIND11_ENABLE_TP_META_CALL_INIT_PROTECTION
#else
#define PYBIND11_ENABLE_SAFE_TP_INIT_INIT_PROTECTION
#endif
```

and wrap the two protection mechanisms in their corresponding conditions??

Reply:
No: Basically, the starting point for the original change in PyCLIF was exactly to find a trick that doesn't involve manipulating the metaclass.

I also thought a little bit about "intercepting" `PyType_Type.tp_call`, but that would be very intrusive and have a runtime impact for pretty much *all* calls creating a Python object.

Reply:
> Maybe defining a few secondary compilation conditions,

Done, thanks for the suggestion! I chose different names for the macros and made it so that the choice of implementation can be overridden externally.